### ISSUE TITLE
[Security Review] WasmEdge

### ISSUE THREAD

**Issue by dm4** [cncf/tag-security#1337]
The WasmEdge team would like to initiate the CNCF TAG-Security Security Assessment (TSSA) process.

Project Name: WasmEdge
Github URL: https://github.com/WasmEdge/WasmEdge
CNCF project stage and issue: https://github.com/cncf/toc/issues/1316 (sandbox)
Security Provider: No

- [ ] Identify team
    - [ ] Project security lead
    - [ ] Lead security reviewer
    - [ ] 1 or more additional reviewer(s)
    - [ ] Every reviewer has read [security reviewer guidelines](/community/assessments/guide/joint-assessment.md) and stated declaration of conflict
    - [ ] Sign off by facilitator on reviewer conflicts
- [x] Create slack channel (e.g. #sec-assess-projectname)
- [ ] Project lead provides draft document - see [outline](/community/assessments/guide/joint-assessment.md)
- [ ] "Naive question phase" Lead Security Reviewer asks clarifying questions
- [ ] Assign issue to security reviewers
- [ ] Initial review
- [ ] Presentation & discussion
- [ ] Share draft findings with project
- [ ] Assessment summary and doc checked into /assessments/projects/project-name (require at least 1 co-chair approval)
- [ ] CNCF TOC presentation (if requested by TOC)

**Comment by JustinCappos** [comment]
We have three other assessments ongoing.  We'll likely have the bandwidth for this in early September.

Who will be the project lead from your side?

**Comment by hydai** [comment]
Hi @JustinCappos 
I am a WasmEdge maintainer[1], and I am going to be the project lead.
Since dm4 is OOO this week, I am submitting the self-assessment instead.

[1]: https://github.com/WasmEdge/WasmEdge/blob/master/docs/OWNER.md#wasmedge-maintainers

**Comment by hydai** [comment]
Hi @JustinCappos 

Is there anything else I should do before you start the review process, such as rebasing the PR?

**Comment by JustinCappos** [comment]
Where is the self assessment document at?   You can update this issue to include a link.

**Comment by hydai** [comment]
> Where is the self assessment document at? You can update this issue to include a link.

Here is the PR: https://github.com/cncf/tag-security/pull/1343/files
Please feel free to let me know if you need more materials for the review process. Thanks.
Cc @dm4 Please check the `draft document` item.

**Comment by JustinCappos** [comment]
Okay, thanks.  I updated the issue description.

Depending on who the assessment lead is, they may prefer that you have
something in a google doc so that it is easier for folks to iterate quickly
on comments.   If so, don't worry at all about the formatting, etc.  Just
get the content there so people can work on it.

We have the assessment of dragonfly on-going.  I'll try to get a team
together for your project while that is on-going.   Certainly, we should be
able to start when that finishes or stalls, if nothing else.

**Comment by dm4** [comment]
I just created a Google Doc to help us collaborate and edit files more easily. Thank you for your help.

https://docs.google.com/document/d/1Mv2AZRwkJjsjoputCyg_IuPk5gfFhQJuCx2fVigkic0/edit

**Comment by dm4** [comment]
Just checking in to see if there are any updates on the security review for WasmEdge. Please let us know if there's anything else we can provide to help with the process. Thank you for your assistance and support.

**Comment by JustinCappos** [comment]
Okay, we need to recruit a group to do this assessment.   We have an assessment ongoing currently, OSS Summit Japan is happening now and KubeCon NA is in 2 weeks.   I think most likely this will not happen until after KubeCon NA.   Sorry for the delay!

**Comment by victorjunlu** [comment]
I am interested as a participant.

**Comment by JustinCappos** [comment]
Great, @victorjunlu .   Would you please read the [guidelines](https://github.com/cncf/tag-security/tree/main/community/assessments/guide/security-reviewer.md) and assert if you have a conflict?

**Comment by JustinCappos** [comment]
I'd also like to volunteer to be a security reviewer.   I have no hard or soft conflicts.

**Comment by victorjunlu** [comment]
@JustinCappos Yes, I assert that I have no hard or soft conflicts. Thanks

**Comment by JustinCappos** [comment]
@brandtkeller @mnm678 @guilhermocc all expressed interest in being a reviewer.   Can you each please read the guidelines and assert if you have any hard or soft conflicts?

**Comment by alabulei1** [comment]
Hi, @JustinCappos @victorjunlu @brandtkeller @mnm678 @guilhermocc  Thank you all for your interest! It would be fantastic to have you as reviewers for WasmEdge. Feel free to let me know if you have any questions about WasmEdge.

**Comment by JustinCappos** [comment]
> [@brandtkeller](https://github.com/brandtkeller) [@mnm678](https://github.com/mnm678) [@guilhermocc](https://github.com/guilhermocc) all expressed interest in being a reviewer. Can you each please read the guidelines and assert if you have any hard or soft conflicts?

@brandtkeller @mnm678 @guilhermocc Hope everyone made it back from KubeCon safely!   When you get a chance, please read the [guidelines](https://github.com/cncf/tag-security/tree/main/community/assessments/guide/security-reviewer.md) and assert if you have a conflict?

**Comment by mnm678** [comment]
I have no hard or soft conflicts.

**Comment by entlein** [comment]
Hi @JustinCappos , I m happy to help in the review, too. I read the guidelines and have no conflicts whatsoever.

**Comment by mrcdb** [comment]
I am happy to lead this security review. I read the guidelines and I have neither hard or soft conflicts.

**Comment by JustinCappos** [comment]
Okay, thanks!   @mrcdb we're ready to go with the naïve questions phase!

**Comment by hydai** [comment]
It should be fixed now. Please check again. Thanks.

**Comment by mrcdb** [comment]
Thanks! A minor fix would be to remove the first section that mirrors the github issue description but other than that it looks good.

FYI I have requested edit access to the document, as the current permissions don't allow me to add comments.

**Comment by camilaavilarinho** [comment]
If possible I’d like to participate as an observer, since this would be my first security assessment here. I have no hard or soft conflicts.

**Comment by matthewflannery** [comment]
I'd like to participate as an observer, I have no conflicts.

**Comment by hydai** [comment]
The first section is removed. I also updated the default permissions; it should be fine to add comments on the documents now. If you need, I can change the default permissions from commenter to editor. Thanks.

**Comment by JustinCappos** [comment]
@camilaavilarinho @matthewflannery Okay, added you both!

**Comment by mrcdb** [comment]
Hi @hydai, sorry for taking some time to update the issue but I was travelling last week.

I have done an initial pass on the self-assessment GDoc as part of the [naive questions phase](https://github.com/cncf/tag-security/blob/main/community/assessments/guide/README.md#clarifying-questions-phase), please have a look at your convenience and address the comments. It is important that comments are addressed in the document text with your clarifications/updates, as the comments (and any reply) will be lost once the assessment is later converted to markdown.

**Comment by dm4** [comment]
Hi @mrcdb,

@hydai and I have made revisions and additions to the document according to the comments. We have also responded to some of the questions mentioned in the comments. Please take a look at your convenience, and let us know if there is anything else we need to address. Thank you!

**Comment by mrcdb** [comment]
Hey @dm4, thanks for the update! I will review the changes soon so we can move forward with the next phase of the assessment with the rest of the review team.

**Comment by mrcdb** [comment]
hey @hydai @dm4 , sorry for the delay on any assessment update (largely due to holidays)

I have reviewed the answers to the clarifying questions and approved all comments, thanks for addressing them in the document text as requested! We can now move to the next phase of the review 🎉

I have assigned the issue to security reviewers @victorjunlu  @JustinCappos  @mnm678 @entlein (tagging @camilaavilarinho @matthewflannery for visibility) for the next [security assessment](https://github.com/cncf/tag-security/tree/main/community/assessments/guide#security-assessment) phase, where each reviewer is tasked with reviewing the self-assessment doc. Ideally, it would be great if we could do this in the next couple of weeks if possible.

**Comment by mrcdb** [comment]
@victorjunlu @JustinCappos @mnm678 @entlein did you all had a chance already to review the [draft document](https://docs.google.com/document/d/1Mv2AZRwkJjsjoputCyg_IuPk5gfFhQJuCx2fVigkic0/edit) ? Just asking as I haven't seen many comments in the document so far so I wonder if you still need some time to provide your reviews.

**Comment by JustinCappos** [comment]
Looks like we've rectified this a bit.   We're surprisingly aligned with
one of the big asks from their team (more detail about the system call
restrictions for the environment.   My sense is that we may need them to
complete that before we're ready to meet and discuss.

**Comment by dm4** [comment]
Thank you all for your reviews. We are currently addressing some questions in the Google document. Regarding system calls and WASI (WebAssembly System Interface), we acknowledge that we didn't explain them clearly enough in the original document. We will add a new section in the background to provide a more detailed explanation.

**Comment by dm4** [comment]
It seems that we have responded to the questions on the Google Doc. If there are any unclear parts that need further explanation, please feel free to ask. Additionally, if there is anything that needs to be completed before moving on to the next phase, please let us know. Thank you!

**Comment by entlein** [comment]
Dear @dm4 , thank you for your answers to the remarks and edits. I answered
the points where I believe your project may achieve better scalability with
(good) tests in the future. But, I dont consider them as blockers.
I would simply state that the easier you make it for outsiders to see the
security and quality of your project without having to dig deep: the more
trust you ll generate and your contributors will have an easier time to
communicate on the importance of tests and relevance of any oversights. in
short: it ll be easier to catch relevant bugs :)
Hope it helps and all the best from me
Cheers
entlein

**Comment by hydai** [comment]
Hi everyone,  
I would like to know if there is anything we need to resolve in the document to move forward with this progress. Or is there any blocking issue we need to fix before the next step?  
If so, please let @dm4 and me know. Thanks!

**Comment by JustinCappos** [comment]
> [@JustinCappos](https://github.com/JustinCappos) Do you think OSPS Baseline compliance should be checked https://baseline.openssf.org/ ?

This is a good thought, but it's not required as part of this process.   In general, if we see a recommendation there which is really relevant, it's fair to bring up.   Note, this would be true of CRA, NIST 800-53, OpenSSF Best Practices Badge, etc.   If something is super relevant, we can mention it, but let's not overly care about conformance with any specific other ruleset.  At this time we just require projects to go through our joint-assessment process.   If the TOC wants to add other requirements, they need to do so (not us).

**Comment by eddie-knight** [comment]
+1 ... The Baseline is currently an optional set of recommendations. Also, the Baseline metrics are typically different from the types of things that a joint assessment looks at — baseline is focused on security hygiene while the assessment process looks at secure design/development.

**Comment by dm4** [comment]
Thanks to all the reviewers for their comments. I have compiled all the comments and responses into the markdown document and submitted pull request #1474.

**Comment by JustinCappos** [comment]
> [@mnm678](https://github.com/mnm678) [@victorjunlu](https://github.com/victorjunlu) [@entlein](https://github.com/entlein) [@camilaavilarinho](https://github.com/camilaavilarinho) [@matthewflannery](https://github.com/matthewflannery) Okay, reviewers, it's time to re-read through the self assessment draft, add any comments, and come up with your thoughts for our part of the assessment. Let's plan to meet with their team in 1 week to share results. So please make a pass by Monday, if possible. (Kindly thumbs up this message to let me know you've seen this)

I didn't get thumbs up from anyone but did get a proof-of-life response from @camilaavilarinho and @mnm678   I did not hear anything from others.   

I'm at KubeCon Japan this week, but perhaps we can talk next week at some point.   I'll be at an OpenSSF event there, but will at least be in a reasonable timeframe.   Please any reviewer who is ready to discuss, please mention this in the issue.   I am ready!

**Comment by entlein** [comment]
I went through the doc and formatted all the links (so that people would have some visual feedback), and I thumbed up your slack message. It is strange that you cant see it. Ready to move ahead, yes

**Comment by JustinCappos** [comment]
> I went through the doc and formatted all the links (so that people would have some visual feedback), and I thumbed up your slack message. It is strange that you cant see it. Ready to move ahead, yes

Sorry, this is likely my fault.   I got confused about who entlein / Duck is.  😆

**Comment by entlein** [comment]
Here's another rfc for the remote workload attestation question we discussed @victorjunlu 
I couldn't find if wasm - in general - has a stance on how to integrate with machine-identity https://github.com/confidential-containers/trustee/issues/812
(In addition to aforementioned https://wasmcloud.com/blog/2025-03-04-why-were-adopting-spiffe-for-webassembly-workload-identity/ )

Does anyone know if there's a rfc for wasm and workload identity overall ?

**Comment by JustinCappos** [comment]
I've posted a doodle poll in the slack channel to plan the discussion with the project.

**Comment by JustinCappos** [comment]
Meeting scheduled for Monday at 9AM NYC time.   Please find the link in the slack channel

**Comment by entlein** [comment]
Thats great, many thanks, feel free to tag me - esp in the formal verification/test-coverage issues.

**Comment by eddie-knight** [comment]
No need to do another issue!

The presentation would be to walk through the assessment content and share a bit about the process. 

It would ideally be 10-30 minutes, but shouldn't require a large amount of preparation.

**Comment by hydai** [comment]
Hi @eddie-knight,  
Thanks for the information. We would like to present the assessment content. Due to our time zone, we prefer to attend the EMEA session. Could you please let us know when the next meeting will be so we can prepare accordingly?

**Comment by JustinCappos** [comment]
> Hi [@eddie-knight](https://github.com/eddie-knight) Thanks for the information. Could we do the presentation this Thursday (APAC), July 17, 2025?

There is already an agenda item this week for APAC.  You might try an EMEA meeting or choose a later APAC slot.

**Comment by hydai** [comment]
Good to know. :-)
We would like to do the presentation in on (AMER) July 24.
Do we need to press the register button on the calendar, or just click the link to join the meeting at that time?

**Comment by eddie-knight** [comment]
Alright, I've added you to the calendar!

Click to join should work fine, I think... not entirely sure. If not, then just do the extra clicks all at once and you'll be good.
